service/matching/tasklist/task_reader.go

```diff
@@ -467,121 +491,7 @@ func (tr *taskReader) dispatchSingleTaskFromBuffer(isolationGroup string, taskIn
-		// it will get a new isolation group to be placed. If it needs re-routing, then
-		// this will be the new routing destination.
-		group, err := tr.getIsolationGroupForTask(tr.cancelCtx, taskInfo)
-		if err != nil {
-			// it only errors when the tasklist is a sticky tasklist and
-			// the sticky pollers are not available, in this case, we just complete the task
-			// and let the decision get timed out and rescheduled to non-sticky tasklist
-			if err == _stickyPollerUnavailableError {
-				tr.completeTask(taskInfo, nil)
-				e.EventName = "Dispatch Failed because StickyPollerUnavailable"
-				event.Log(e)
-				return false, true
-			}
-			// it should never happen, unless there is a bug in 'getIsolationGroupForTask' method
-			tr.logger.Error("taskReader: unexpected error getting isolation group",
-				tag.Error(err),
-				tag.IsolationGroup(group))
-
-			e.EventName = "Dispatch Failed due to unexpected error getting isolation group"
-			e.Payload = map[string]any{
-				"error": err,
-			}
-			event.Log(e)
-			e.Payload = nil
-
-			tr.completeTask(taskInfo, err)
-			return false, true
-		}
-
-		if group == isolationGroup {
-			// no change, retry to dispatch the task again
-			return false, false
-		}
-
-		// ensure the isolation group is configured and available
-		_, taskGroupReaderIsPresent := tr.taskBuffers[group]
-		if !taskGroupReaderIsPresent {
-			// there's a programmatic error. Something has gone wrong with tasklist instantiation
-			// don't block and redirect to the default group
-			tr.scope.IncCounter(metrics.BufferIsolationGroupRedirectFailureCounter)
-			tr.logger.Error("An isolation group buffer was misconfigured and couldn't be found. Redirecting to default",
-				tag.Dynamic("redirection-from-isolation-group", isolationGroup),
-				tag.Dynamic("redirection-to-isolation-group", group),
-				tag.IsolationGroup(group),
-				tag.WorkflowRunID(taskInfo.RunID),
-				tag.WorkflowID(taskInfo.WorkflowID),
-				tag.TaskID(taskInfo.TaskID),
-				tag.WorkflowDomainID(taskInfo.DomainID),
-			)
-
-			select {
-			case <-tr.cancelCtx.Done():
-				// the task reader is shutting down
-				e.EventName = "Dispatch Failed because task reader is shutting down"
-				event.Log(e)
-				return true, true
-			case tr.taskBuffers[defaultTaskBufferIsolationGroup] <- taskInfo:
-				// task successfully rerouted to default tasklist
-				e.EventName = "Task rerouted to default isolation group"
-				event.Log(e)
-				return false, true
-			default:
-				// couldn't redirect, loop and try again
-				e.EventName = "Task is not rerouted to default isolation group. Will retry dispatch"
-				event.Log(e)
-				return false, false
-			}
-		}
-
-		// if there is no poller in the isolation group or the isolation group is drained,
-		// we want to redistribute the tasks to other isolation groups in this case to drain
-		// the backlog.
-		select {
-		case <-tr.cancelCtx.Done():
```

Comment:
Super happy to see this go, the level of complexity here was far too high to reason about.